### PR TITLE
Opt proxy protocol

### PR DESCRIPTION
**Pull request by sdvdxl**
Signed-off-by: sdvdxl sdvdxl@163.com

Motivation:

Vertx supports proxy settings and override options methods to facilitate calls. #152

### CODE REVIEW

**Comment by ppatierno** [comment]
Do you think it's possible to write a test for this?

**Comment by vietj** [comment]
there is in vertx-tests an utility class to setup a fake minimalist proxy for this https://github.com/eclipse-vertx/vert.x/blob/master/src/test/java/io/vertx/test/proxy/HAProxy.java that can be reused

**Comment by vietj** [comment]
well it's a true proxy :-)

**Comment by sdvdxl** [comment]
> Do you think it's possible to write a test for this?

I add 2 tests. V1_TCP4 and V2_TCP4

File: pom.xml [hunk 1]
```diff
@@ -63,6 +63,10 @@
       <groupId>io.netty</groupId>
       <artifactId>netty-codec-mqtt</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-codec-haproxy</artifactId>
```
Comment (vietj) [inline pom.xml:68]: can you set this dependency to optional ?

**Comment by vietj** [comment]
can you add ref the example12 in the `index.adoc` file ?

File: pom.xml (hunk 1 above)
Comment (ppatierno) [reply pom.xml:68]: Do you mean `test` scope?

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 2]
```diff
@@ -44,14 +44,21 @@ public void example1(Vertx vertx) {
...
+            .getPassword() + "]");
       }
       if (endpoint.will() != null) {
-        System.out.println("[will topic = " + endpoint.will().getWillTopic() + " msg = " + new String(endpoint.will().getWillMessageBytes()) +
-          " QoS = " + endpoint.will().getWillQos() + " isRetain = " + endpoint.will().isWillRetain() + "]");
+        System.out.println(
+          "[will topic = " + endpoint.will().getWillTopic() + " msg = " + new String(
+            endpoint.will().getWillMessageBytes()) +
+            " QoS = " + endpoint.will().getWillQos() + " isRetain = " + endpoint.will()
+            .isWillRetain() + "]");
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:61]: why these changes?

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 3]
```diff
@@ -103,14 +112,21 @@ public void example3(Vertx vertx) {
...
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:129]: ditto as above

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 4]
```diff
@@ -267,11 +293,44 @@ public void example10(Vertx vertx) {
...
+  public void example12(Vertx vertx) {
+    MqttServer mqttServer = MqttServer
+      .create(vertx, new MqttServerOptions()
+        // set true to use proxy protocol
+        .setUseProxyProtocol(true));
+    mqttServer.endpointHandler(endpoint -> {
+      // remote address is origin real address， not proxy's address
+      System.out.println(endpoint.remoteAddress());
+      endpoint.accept(false);
+
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:322]: empty line

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 5]
```diff
@@ -267,11 +293,44 @@ public void example10(Vertx vertx) {
...
+    })
+      .listen(ar -> {
+
+        if (ar.succeeded()) {
+
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:327]: empty line

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 6]
```diff
@@ -267,11 +293,44 @@ public void example10(Vertx vertx) {
...
+          System.out.println("MQTT server is listening on port " + ar.result().actualPort());
+        } else {
+
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:330]: empty line

File: src/main/java/examples/VertxMqttServerExamples.java [hunk 7]
```diff
@@ -267,11 +293,44 @@ public void example10(Vertx vertx) {
...
+      .create(vertx, new MqttServerOptions()
+        // set true to use proxy protocol
+        .setUseProxyProtocol(true));
+    mqttServer.endpointHandler(endpoint -> {
+      // remote address is origin real address， not proxy's address
+      System.out.println(endpoint.remoteAddress());
+      endpoint.accept(false);
+
+    })
+      .listen(ar -> {
```
Comment (ppatierno) [inline src/main/java/examples/VertxMqttServerExamples.java:324]: can you make this call on the same line as the previous parenthesis?

File: src/test/java/io/vertx/mqtt/test/server/MqttServerWithProxyConnectionTest.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    MqttServerOptions options = new MqttServerOptions().setUseProxyProtocol(true);
+    options.setTimeoutOnConnect(MQTT_TIMEOUT_ON_CONNECT);
+
+    this.setUp(context, options);
+  }
+
+  private void createHaproxy(HaproxyVersion version) throws Exception {
+    haProxy = new HAProxy(
+      SocketAddress.inetSocketAddress(MQTT_SERVER_PORT, MQTT_SERVER_HOST), version.header);
+    haProxy.start(Vertx.vertx());
```
Comment (ppatierno) [inline src/test/java/io/vertx/mqtt/test/server/MqttServerWithProxyConnectionTest.java:67]: why creating a new Vert.x instance, can't you use the one coming from the `MqttServerBaseTest` class?

File: src/main/java/examples/VertxMqttServerExamples.java (hunk 2 above)
Comment (vietj) [reply src/main/java/examples/VertxMqttServerExamples.java:61]: yes test scope would be fine too. Vert.x Core defines an optional dependency onto this artifact because it needs to compile with it.

File: pom.xml (hunk 1 above)
Comment (vietj) [reply pom.xml:68]: test scope would be fine too.

Vert.x Core defines an optional dependency onto this artifact because it needs to compile with it.

File: src/main/java/examples/VertxMqttServerExamples.java (hunk 2 above)
Comment (sdvdxl) [reply src/main/java/examples/VertxMqttServerExamples.java:61]: jetbrains idea format result. I will change it.

File: src/main/java/examples/VertxMqttServerExamples.java (hunk 4 above)
Comment (sdvdxl) [reply src/main/java/examples/VertxMqttServerExamples.java:322]: well, This code turns out to be like this. Did you mean you want to remove this empty line now?

File: src/main/java/examples/VertxMqttServerExamples.java (hunk 7 above)
Comment (sdvdxl) [reply src/main/java/examples/VertxMqttServerExamples.java]: This turned out to be the case, I thought it was a code template.

**Comment by ppatierno** [comment]
@vietj why Travis is complaining with following ...

```shell
io.vertx.docgen.DocGenException: Could not resolve examples.VertxMqttClientExamples#example12
```

This is a server example not a client one.

**Comment by vietj** [comment]
changes the ref in the asciidoc file

File: src/main/asciidoc/index.adoc [hunk 9]
```diff
@@ -308,3 +308,32 @@ WARNING: The handler WILL NOT BE CALLED if sent publish packet with QoS=0.
 ----
 {@link examples.VertxMqttClientExamples#example9}
 ----
+
+=== Use proxy protocol
+
+[source,$lang]
+----
+{@link examples.VertxMqttClientExamples#example12}
```
Comment (ppatierno) [inline src/main/asciidoc/index.adoc:316]: @sdvdxl this is the problem as pointed by @vietj

Comment (sdvdxl) [reply src/main/asciidoc/index.adoc]: fixed

**Comment by ppatierno** [comment]
@vietj I don't have rights for restart Travis build, can you do that?

**Comment by vietj** [comment]
done

**Comment by ppatierno** [comment]
@sdvdxl Travis is still failing. Are you able to run the entire tests suite locally?

**Comment by sdvdxl** [comment]
> @sdvdxl Travis is still failing. Are you able to run the entire tests suite locally?

@ppatierno The local tests passed, but on my other remote machine it was stuck at `MqttServerMaxMessageSizeTest` and could no longer be executed

**Comment by sdvdxl** [comment]
I think the issue maybe is line 71 of `MqttServerMaxMessageSizeTest:      client.publish(MQTT_TOPIC, message, 0, false);`. 

The message over the max bytes, and the sever throw a exception, the client wait forever(client 
 default timeout is -1)

**Comment by sdvdxl** [comment]
Before  `client.connect();` I set `client.setTimeToWait(TimeUnit.SECONDS.toMillis(3));`
after running a few times(sometime passed, sometime failed) then print:

```java
Timed out waiting for a response from the server (32000)
	at org.eclipse.paho.client.mqttv3.internal.Token.waitForCompletion(Token.java:112)
	at org.eclipse.paho.client.mqttv3.MqttToken.waitForCompletion(MqttToken.java:67)
	at org.eclipse.paho.client.mqttv3.MqttClient.publish(MqttClient.java:399)
```

See this pr [Fix MQTT client hangs after publishing with QoS 0](https://github.com/aykutaras/paho.mqtt.java/pull/3) , we need to upgrade our paho version to fix it.

**Comment by ppatierno** [comment]
@sdvdxl tbh I am not sure how this fix the "problem". You are waiting asynchronously during a connection before publishing the message. The synchronous connection is used on all over the other tests, right? And it works fine.

**Comment by vietj** [comment]
what is the status on this issue @ppatierno ?

**Comment by ppatierno** [comment]
@vietj I raised a problem but I don't see any changes from @sdvdxl

**Comment by vietj** [comment]
@sdvdxl ping we would like to merge this PR
